toolchain/check/handle_function.cpp

```diff
@@ -60,6 +61,55 @@ static auto DiagnoseModifiers(Context& context, bool is_definition,
   return context.decl_state_stack().innermost().modifier_set;
 }
 
+// Returns the function ID for the instruction, or invalid.
+static auto GetRedeclFunctionId(Context& context, SemIR::Inst prev_inst)
+    -> SemIR::FunctionId {
+  CARBON_KIND_SWITCH(prev_inst) {
+    case CARBON_KIND(SemIR::StructValue struct_value): {
+      if (auto fn_type = context.types().TryGetAs<SemIR::FunctionType>(
+              struct_value.type_id)) {
+        return fn_type->function_id;
+      }
+      return SemIR::FunctionId::Invalid;
+    }
```

Comment:
Do we still need this case at all? I have a suspicion that we stopped needing it as of #3917.

Reply:
Yes, because this is the constant value.

Reply:
Sure, but shouldn't `prev_inst` here be the previous instruction found by name lookup? I would expect that to be a `FunctionDecl`, not a `StructValue`.

In any case, consider this non-blocking: this PR is preserving the existing behavior here.

Reply:
That's not how it works. As a reminder, this is pretty much the same issue that came up when I worked on `class`, discussed here: https://github.com/carbon-language/carbon-lang/pull/3893#discussion_r1574889278

Reply:
Yeah, thanks. So the issue is that name lookup will find an `ImportRef`, not a `FunctionDecl`, and we don't have a good way to determine if the `ImportRef` actually refers to a `FunctionDecl` because all we have is the constant value. That makes sense, and while it's not going to be correct in the long term, that's an issue for another day.